### PR TITLE
Fix py http parser not treating 204/304/1xx as an empty body

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

There was a disagreement on how to handle 204/304/1xx responses between the c parser and py parser.

204/304/1xx are now always treated as an empty body in the py parser to match the c parser and comply with
https://datatracker.ietf.org/doc/html/rfc9112#section-6.3

## Are there changes in behavior for the user?

A empty chunked response body `0\r\n\r\n` will no longer be read for 204, 203, 1xx responses when using the py parser. This matches the behavior of the c parser.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

TODO: 

- [x] timeline
- [x] mypy
- [x] more functional testing

### CODE REVIEW

File: aiohttp/http_parser.py [hunk 1]
```diff
@@ -338,12 +338,19 @@ def get_content_length() -> Optional[int]:
...
                         # calculate payload
+                        # 204, 304, 1xx should not have a body per
+                        # https://datatracker.ietf.org/doc/html/rfc9112#section-6.3
+                        code_indicates_empty_body = (
+                            code in (204, 304) or 100 <= code < 200
+                        )
                         if (
                             (length is not None and length > 0)
-                            or msg.chunked
+                            or (msg.chunked and not code_indicates_empty_body)
```
Comment (Dreamsorcerer) [inline aiohttp/http_parser.py]: Shouldn't this override it even if length>0? i.e. It shouldn't be bracketed.

File: tests/test_http_parser.py [hunk 2]
```diff
@@ -1060,6 +1061,131 @@ def test_parse_no_length_payload(parser: Any) -> None:
     assert payload.is_eof()
 
 
+def test_parse_content_length_payload_multiple(response: Any) -> None:
+    text = b"HTTP/1.1 200 OK\r\n" b"content-length: 5\r\n\r\n" b"first"
```
Comment (Dreamsorcerer) [inline tests/test_http_parser.py]:
```suggestion
    text = b"HTTP/1.1 200 OK\r\ncontent-length: 5\r\n\r\nfirst"
```

Comment (Dreamsorcerer) [reply tests/test_http_parser.py]: And same for all the other ones...

Comment (bdraco) [reply tests/test_http_parser.py]: Adjusted to collapse string

File: aiohttp/http_parser.py (hunk 1 above)
Comment (bdraco) [reply aiohttp/http_parser.py]: Refactored to be included with the length check as well

**Comment by bdraco** [comment]
https://github.com/aio-libs/aiohttp/issues/6446 is a bit of a problem since its a special case

**Comment by bdraco** [comment]
the `msg.upgrade` check looks like it might have a problem

**Comment by bdraco** [comment]
not 100% sure if the msg.upgrade check it supposed to apply to chunk and content length or just chunked given https://github.com/aio-libs/aiohttp/commit/6301d6b75f86f31096fb0212c57238b4604f02a1

**Comment by bdraco** [comment]
I'm not sure its right but I made it `or (msg.chunked and not msg.upgrade)` so the check is the same.

If its not right it should probably be fixed in another PR as I didn't want to break compatibility.

**Comment by bdraco** [comment]
I think this one is ready for another look.

If it looks good I'll start doing production testing

I'm still working through regressions in the other one

File: aiohttp/http_parser.py [hunk 3]
```diff
@@ -383,34 +390,33 @@ def get_content_length() -> Optional[int]:
                                 auto_decompress=self._auto_decompress,
                                 lax=self.lax,
                             )
+                        elif (
+                            not code_indicates_empty_body
```
Comment (Dreamsorcerer) [inline aiohttp/http_parser.py]: I'm thinking in both PRs, maybe `must_be_empty_body` feels a little clearer to me when reading the if statements.

**Review by Dreamsorcerer — APPROVED** [review]

File: aiohttp/http_parser.py (hunk 3 above)
Comment (bdraco) [reply aiohttp/http_parser.py]: adjusted in 68c5bc2fbceb39a04c7b7abeaa4509bdac989eea

File: aiohttp/http_parser.py [hunk 4]
```diff
@@ -338,10 +338,19 @@ def get_content_length() -> Optional[int]:
...
                         assert self.protocol is not None
                         # calculate payload
-                        if (
+                        # 204, 304, 1xx should not have a body per
+                        # https://datatracker.ietf.org/doc/html/rfc9112#section-6.3
+                        must_be_empty_body = (
+                            method in (hdrs.METH_CONNECT, hdrs.METH_HEAD)
+                            or code in (204, 304)
+                            or 100 <= code < 200
+                        )
```
Comment (bdraco) [inline aiohttp/http_parser.py]: Maybe move this to helpers.py so its a bit more DRY

File: aiohttp/client.py [hunk 5]
```diff
@@ -526,7 +527,7 @@ async def _request(
                     assert conn.protocol is not None
                     conn.protocol.set_response_params(
                         timer=timer,
-                        skip_payload=method.upper() == "HEAD",
+                        skip_payload=method_must_be_empty_body(method.upper()),
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]: Probably a good idea to put the .upper() in the helper?

**Comment by Dreamsorcerer** [comment]
This is still marked as a draft, so remember to update that if you think it's ready to merge.

File: aiohttp/client.py (hunk 5 above)
Comment (bdraco) [reply aiohttp/client.py]: adjusted in d1fb2c103a9be5683a93e3106e06247bca0b3c5e

**Comment by bdraco** [comment]
> This is still marked as a draft, so remember to update that if you think it's ready to merge.

Will do.  Still doing functional testing, and need to write the timeline message

**Pull request merged**